beluga/include/beluga/algorithm/unscented_transform.hpp

```diff
@@ -0,0 +1,107 @@
+// Copyright 2024 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BELUGA_ALGORITHM_UNSCENTED_TRANSFORM_HPP
+#define BELUGA_ALGORITHM_UNSCENTED_TRANSFORM_HPP
+
+#include <Eigen/Cholesky>
+#include <Eigen/Core>
+
+#include <optional>
+#include <range/v3/range/concepts.hpp>
+#include <range/v3/range/conversion.hpp>
+#include <range/v3/view/any_view.hpp>
+#include <range/v3/view/common.hpp>
+#include <range/v3/view/transform.hpp>
+#include <range/v3/view/zip.hpp>
+#include <type_traits>
+#include <utility>
+#include <vector>
+#include "beluga/eigen_compatibility.hpp"
+namespace beluga {
+
+/**
+ * \brief Implements the unscented transform, a mathematical function used to estimate the result of applying a given
+ *  a possibly nonlinear transformation to a probability distribution that is characterized with mean and covariance.
+ * See https://en.wikipedia.org/wiki/Unscented_transform for more information.
+ *
+ * \tparam MeanDerived Concrete Eigen dense type used for the mean.
+ * \tparam CovarianceDerived Concrete Eigen dense type used for the covariance.
+ * \tparam TransferFn Callable that maps a vector from input space to output space. Possibly non-linear.
+ * \param mean Mean in the input space.
+ * \param covariance Covariance in the input space.
+ * \param transfer_fn Callable so that
+ * \param kappa Parameter used for sigma points selection. See https://arxiv.org/pdf/2104.01958 Equation (5). A sensible
+ * default will be used if not provided.
+ * \return A pair containing (mean, estimate) for the output space.
+ */
+template <typename MeanDerived, typename CovarianceDerived, typename TransferFn>
+auto unscented_transform(
+    const Eigen::MatrixBase<MeanDerived>& mean,
+    const Eigen::MatrixBase<CovarianceDerived>& covariance,
+    TransferFn&& transfer_fn,
+    std::optional<typename MeanDerived::Scalar> kappa = std::nullopt) {
+  using Scalar = typename MeanDerived::Scalar;
+  static_assert(
+      std::is_same_v<typename MeanDerived::Scalar, typename CovarianceDerived::Scalar>,
+      "Mean and covariance scalar types differ.");
+  static_assert(std::is_floating_point_v<Scalar>);
+  static_assert(MeanDerived::ColsAtCompileTime == 1, "Mean should be a column vector.");
```

Comment:
@serraramiro1 nit: worth asserting that `CovarianceDerived` is square and that it matches the dimension of `MeanDerived`.

Reply:
Done